### ISSUE TITLE
 is there a way to restore and predict without doing "dummy training"  

### ISSUE THREAD

**Issue by satendrapratap** [tensorflow/tensorflow#13971]
We are trying **text_classification.py** example of tensorflow  and separated the training and prediction parts of the code and trained the model using model_dir and tried to predict using the same model_dir.

Model is being saved but when we comment the training code and try to predict something, we get the following error:

**InvalidArgumentError (see above for traceback): Assign requires shapes of both tensors to match. lhs shape= [10,50] rhs shape= [5386,50]
         [[Node: save/Assign = Assign[T=DT_FLOAT, _class=["loc:@EmbedSequence/embeddings"], use_locking=true, validate_shape=true, _device="/job:localhost/replica:0/task:0/cpu:0"](EmbedSequence/embeddings, save/RestoreV2)]]**

**Where:** 5386 and 10 are words count in training and prediction data.

But when we uncomment the preparation of train_input_fn without using it anywhere in the code , without training the model but using the same model_dir then prediction works fine.

train_input_fn = tf.estimator.inputs.numpy_input_fn(
        x={WORDS_FEATURE: x_train},
        y=y_train,
        batch_size=len(x_train),
        num_epochs=None,
        shuffle=True)

We are wondering what exactly is being done by this function which actually overcome the error without training the model (using the function at all) ?

We have gone through many posts but following post looks more relevant where someone said that prediction can't be done without doing at least some training (though the model was already trained and restored using model_dir ):  https://github.com/tensorflow/tensorflow/issues/3340

------------------------

### System information
- **Tried both Python 2.7.6 / Python 3.4.3**:
- **TF Version :  1.3.0**:
- **python -c "import tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)" :  ('v1.1.0-rc0-61-g1ec6ed5', '1.1.0')**

**Comment by michaelisard** [comment]
This question is better asked on  [StackOverflow](http://stackoverflow.com/questions/tagged/tensorflow) since it is not a  bug or feature request. There is also a larger community that reads questions there. Thanks!

**Comment by satendrapratap** [comment]
As explained in the https://github.com/tensorflow/tensorflow/issues/3340  , it was a to be implemented feature as told by @martinwicke : 
"Right now, there isn't a way to restore and predict without running at least one step of training (which is a missing feature)."

Has it been released ?

**Comment by martinwicke** [comment]
Yes, tf.estimator.* does not have this restriction.

**Comment by satendrapratap** [comment]
[text_classification.txt](https://github.com/tensorflow/tensorflow/files/1421562/text_classification.txt)


We are using TF 1.3.0 and  [text_classification.py]([url](url))  uses tf.estimator but we still face that issue.
You can reproduce the issue by running the attached code (convert .txt to .py).


To train without testing use (no error):
SP_TRAIN1=1
SP_TRAIN=1
SP_TEST=0

To test without training use (you will get the error in this case):
SP_TRAIN1=0
SP_TRAIN=0
SP_TEST=1

To test without training but uncomment training input function, use (you will NOT get the error in this case):
SP_TRAIN1=1
SP_TRAIN=0
SP_TEST=1

**Comment by martinwicke** [comment]
Thank you for the repro code!

@ispirmustafa can you take a look?

**Comment by facaiy** [comment]
I agree with @martinwicke that tf.estimator.* does not have this restriction (at least one step train before predict).

If I understand you correctly, a simple code below might work as you expected:
```python
if Train:
    estimator.train(train_input_fn, xxx)
else:
    estimator.predict(test_input_fn, xxx)
```

note: Do not comment training code anywhere, because they might be used to reconstruct the graph.

**Comment by satendrapratap** [comment]
>> note: Do not comment training code anywhere, because they might be used to reconstruct the graph.

Thats exactly the point that if we are trying to predict on a trained model then why do we need the training code at all?

Below line of code should take care of everything related to reconstructing the graph etc in case of predicting so that we should not any training code during prediction:
classifier = tf.estimator.Estimator(model_fn=model_fn, model_dir=MODEL_DIR)

**Comment by satendrapratap** [comment]
@ispirmustafa  could you please take a look at it?

**Comment by satendrapratap** [comment]
is there any update on this yet ?

**Comment by ispirmustafa** [comment]
Hi @satendrapratap ,
I think this issue is about VocabularyProcessor. if only SP_TEST is set, then there isn't any call to VocabularyProcessor.fit_transform. in SP_TEST you call VocabularyProcessor.transform. that creates a dependency to your train code.
Could you please move VocabularyProcessor related logic outside of train and test functionality?

**Comment by satendrapratap** [comment]
Thats the whole point if we are doing testing then why do we need training code (as the model is already trained and we provided the path for all the files in tf.estimator.Estimator(model_fn=model_fn, model_dir=MODEL_DIR) ) . Please checkout comment "satendrapratap commented on Oct 30 •  edited " above.

**Comment by ispirmustafa** [comment]
Estimator does not need that. but the utility that you're using (VocabularyProcessor) is independent then Estimator. It doesn't save it's state to model-dir. it's in memory training. That's why you have the problem.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
